Automated VMI Workload Component Updates

davidvossel opened the pull request:
# Overview
Today, our control plane will update with zero downtime, however the components running within the VMI pod's environment are not touched. This means that after a KubeVirt update the control plane and any future VMIs will be running the most recent deployed release, however existing VMIs that were started before the KubeVirt update will continue to run out dated components.

In this PR I've added a controller which automatically updates the VMI workload components for existing VMIs. This is primarily achieved using LiveMigration which will spawn a new pod with updated components within it and then live migrate an existing VMI into that new environment. In clusters where live migration isn't possible due to storage or network constraints, a secondary method is to simply automate the shutdown of out of date VMIs via pod eviction. VMIs backed by a VM with the runStrategy set to "always" will immediately get restarted, however standalone VMIs will not come back.

## New KubeVirt Config Options

New API fields on the KubeVirt CRD expose the ability to configure if and how automated workload updates occur. This is all configured using the new WorkloadUpdateStrategy API fields.

For example, the following config will automatically update VMI workloads either by LiveMigrate or Evict. VMI's which are live migratable are guaranteed to be live migrated, VMIs that are not live migratable will have their pods evicted which typically results in a shutdown.

For VMIs that are LiveMigrated, the cluster's global max count for parallel migrations will be observed to limit how many migrations happen at once. For shutdown, we have some new options on the workloadUpdateStrategy field to help with that. In the example, we throttle evicting 10 vmi pods per minute.

```
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
  imagePullPolicy: IfNotPresent
  workloadUpdateStrategy:
    workloadUpdateMethods:
      - LiveMigrate
      - Evict
    batchEvictSize: 10
    batchEvictInterval: "1m"
```

By specifying only the "LiveMigrate" strategy, outdated VMIs that are unable to live migrate will not be touched. By specifying only the "Evict" method, all outdated VMIs whether they are live migrateable or not will have their underlying pods evicted.

## Outdated VMI Label

In order to detect outdated VMIs, the VMI controller now sets a new Label ```kubevirt.io/outdatedLauncherImage``` on each VMI that indicates when a running VMI's pod is not using the most recent container image. This label is used by the automated workload update controller to detect outdated VMIs. It's also helpful for users who may want to create their own logic to handle outdated VMIs external of the automation in this PR. 


## Outdated Workload Counter on KubeVirt CR

As a quick way to reference the number of outdated workloads within a cluster, a new value has been added to the KubeVirt CRD's status called ```outdatedVMIWorkloads```

Example:

```
apiVersion: kubevirt.io/v1
kind: KubeVirt
metadata:
  name: kubevirt
  namespace: kubevirt
spec:
status:
  outdatedVMIWorkloads: 7
```

Those outdated VMIs can be detected using the ```kubevirt.io/outdatedLauncherImage``` label as a label selector when listing all VMIs.

## Monitoring and Alerts

New monitoring metric called `kubevirt_vmi_outdated_count` indicates the total number of outdated VMIs in the cluster at a given point in time.

New alert `OutdatedVMIWorkloads` fires after `kubevirt_vmi_outdated_count` != 0 for longer that 24 hours. 

## Testing
Existing operator functional tests have been extended to validate automated workload updates occur. Since workload updates exercise migrations, this has replaced the explicitly defined migration logic in the `update from previous release` functional test. 

## Release Note

```release-note
Automated workload updates via new KubeVirt WorkloadUpdateStrategy API
```

davidvossel commented:
/test pull-kubevirt-unit-test

davidvossel commented:
/retest

fabiand commented:
An incomplete review, but a question: Why "shutdown" and not "reschedule"/"restart"?

davidvossel commented:
> An incomplete review, but a question: Why "shutdown" and not "reschedule"/"restart"?

the update controller works on VMIs not VMs. the action taken on a VMI is shutdown or migrate. shutdown may or may not resutl in reschedule/restart depending on if the VMI is backed by a VM.

davidvossel commented:
This PR should be ready to go. I finished unit tests and added monitoring/alerting

davidvossel commented:
/retest

enp0s3 commented:
/retest

enp0s3 reviewed (changes requested):
Great work! just found a small typo, please see below

rmohr commented on pkg/virt-controller/watch/workload-updater/workload-updater.go:
```diff
@@ -0,0 +1,554 @@
+
+func (c *WorkloadUpdateController) enqueueKubeVirt(obj interface{}) {
+	logger := log.Log
+	kv := obj.(*virtv1.KubeVirt)
```
This cast can fail under certain scenarios. In theory you  can get a tombstone object and you have to do something like this (copied from another watcher):

```
	domain, ok := obj.(*api.Domain)
	if !ok {
		tombstone, ok := obj.(cache.DeletedFinalStateUnknown)
		if !ok {
			log.Log.Reason(fmt.Errorf("couldn't get object from tombstone %+v", obj)).Error("Failed to process delete notification")
			return
		}
		domain, ok = tombstone.Obj.(*api.Domain)
		if !ok {
			log.Log.Reason(fmt.Errorf("tombstone contained object that is not a domain %#v", obj)).Error("Failed to process delete notification")
			return
		}
		return
	}
```

davidvossel replied:
i added logic to simply return early if the kubevirt object can't be cast. there's nothing for this control loop to do if the operator cr is deleted, so i don't see any value in retrieving the tombstone.

rmohr replied:
Ack.

rmohr commented on pkg/virt-controller/watch/workload-updater/workload-updater.go:
```diff
@@ -0,0 +1,554 @@
+		return false
+	}
+
+	_, labelExists := vmi.Labels[virtv1.OutdatedLauncherImageLabel]
```
Since the controller binary knows the target launcher image, maybe just pass it to the controller instead of using this label. Also since we have a shared infromer for VMIs anyway, the label does also not help to do less.

davidvossel replied:
The idea is that the label is meant as a way to list outdated VMIs using a label selector. So, someone can see via monitoring they have `x` number of outdated vmis in the cluster, and then list then using a label selector.

This label would allow someone to create their own vmi update logic if they wanted as well as help field engineers detect which VMIs might be stuck unable to update.

rmohr commented on pkg/virt-controller/watch/workload-updater/workload-updater.go:
```diff
@@ -0,0 +1,554 @@
+	return false
+}
+
+func isMigratable(vmi *virtv1.VirtualMachineInstance) bool {
```
Don't we have a common method for that yet? Maybe work having a common place for this?

davidvossel replied:
oh, we do actually. good catch :+1:

rmohr commented on pkg/virt-controller/watch/workload-updater/workload-updater.go:
```diff
@@ -0,0 +1,554 @@
+	return c
+}
+
+func (c *WorkloadUpdateController) getKubeVirtKey() (string, error) {
```
Don't we have a common method for that yet? Maybe work having a common place for this?

davidvossel replied:
yup, i'll add that

rmohr commented on pkg/virt-controller/watch/workload-updater/workload-updater.go:
```diff
@@ -0,0 +1,554 @@
+		}
+
+		// add label to outdated vmis for sorting
+		data.allOutdatedVMIs = append(data.allOutdatedVMIs, vmi)
```
This line does not seem to add a label or do anything related.

davidvossel replied:
comment removed

rmohr reviewed (changes requested):
I want to do another pass over the logic. Here the initial set of questions.

davidvossel commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -525,6 +526,47 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
+				vmiCopy.Labels[virtv1.OutdatedLauncherImageLabel] = ""
+			} else {
+				if vmiCopy.Labels != nil {
+					delete(vmiCopy.Labels, virtv1.OutdatedLauncherImageLabel)
```
convenience and ability to use label selectors to perform reporting and detection of outdated vmis using standard kubectl tooling.

davidvossel commented on pkg/virt-handler/vm.go:
```diff
@@ -817,6 +819,7 @@ func (d *VirtualMachineController) updateVMIStatus(vmi *v1.VirtualMachineInstanc
 			// At this point we know that the migration has completed and that
 			// the target node has seen the domain event.
 			vmi.Labels[v1.NodeNameLabel] = migrationHost
+			delete(vmi.Labels, v1.OutdatedLauncherImageLabel)
```
it's intentional. see other comments. we can debate the usefulness

davidvossel commented on pkg/virt-operator/util/config.go:
```diff
@@ -129,14 +132,22 @@ func GetConfigFromEnv() (*KubeVirtDeploymentConfig, error) {
 }
 
 func GetTargetConfigFromKV(kv *v1.KubeVirt) *KubeVirtDeploymentConfig {
+	additionalProperties := getKVMapFromSpec(kv.Spec)
```
This map was added to prevent needing to keep adding new fields to this deployment config. this is an internal structure and not an exposed api.

davidvossel commented on manifests/generated/kv-resource.yaml:
```diff
@@ -612,6 +612,22 @@ spec:
+              description: WorkloadUpdateStrategy defines at the cluster level how to handle automated workload updates
+              properties:
+                batchShutdownCount:
+                  description: "BatchShutdownCount Represents the number of VMIs that can be forced updated per the BatchForceInteral interval \n Defaults to 10"
```
works for me, goign with BatchShutdownSize

davidvossel commented:
@enp0s3 @rmohr  all comments should be addressed at this point

davidvossel commented on staging/src/kubevirt.io/client-go/api/v1/types.go:
```diff
@@ -1300,6 +1300,55 @@ type KubeVirtCertificateRotateStrategy struct {
+	WorkloadUpdateMethods []WorkloadUpdateMethod `json:"workloadUpdateMethods,omitempty"`
+
+	// BatchShutdownCount Represents the number of VMIs that can be forced updated per
+	// the BatchForceInteral interval
```
yep, fixed

davidvossel commented on pkg/virt-controller/watch/workload-updater/workload-updater.go:
```diff
@@ -29,14 +30,21 @@ const (
 	SuccessfulCreateVirtualMachineInstanceMigrationReason = "SuccessfulCreate"
+	// FailedDeleteVirtualMachineReason is added in an event if a deletion of a VMI fails
+	FailedDeleteVirtualMachineInstanceReason = "FailedDelete"
+	// SuccessfulDeleteVirtualMachineReason is added in an event if a deletion of a VMI fails
```
fixed, pushing the update now.

davidvossel commented:
@rmohr I added a `vmi.status.launcherContainerImageVersion` field. The workload updater now determines if a vmi is outdated based on comparing that field with the current desired launcher image provided to virt-controller

davidvossel commented on pkg/virt-controller/watch/workload-updater/workload-updater.go:
```diff
@@ -0,0 +1,546 @@
+	for _, vmi := range deletionCandidates {
+		go func(vmi *virtv1.VirtualMachineInstance) {
+			defer wg.Done()
+			err := c.clientset.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &v1.DeleteOptions{})
```
needs to be Evict

davidvossel replied:
done

davidvossel commented:
@rmohr This should be ready for review again.

I changed the behavior to perform a pod eviction rather than deleting the VMI.  The `Shutdown` method has now been called `Evict` and i've changed the corresponding batch related tunings as well to reflect this change.

davidvossel commented:
/retest

davidvossel commented:
/test pull-kubevirt-check-tests-for-flakes

davidvossel commented:
/retest

davidvossel commented:
/retest

rmohr commented on api/openapi-spec/swagger.json:
```diff
@@ -10322,6 +10326,10 @@
      "operatorVersion": {
       "type": "string"
      },
+     "outdatedVMIWorkloads": {
```
Really minor: How about either `outdatedWorkloads` or `outdatedVirtualmachineInstanceWorkloads`? I know it is long but we don't use abbreviations in the API for our our components afaik.

davidvossel replied:
yep, that makes sense to me.

davidvossel replied:
fixed

rmohr reviewed (commented):
/approve

From a high level and API perspective, looks good to me. Did also not see any showstoppers in the controller.

It would be great if someone else can do a deep code review.

rmohr commented on pkg/virt-controller/watch/workload-updater/workload-updater.go:
```diff
@@ -0,0 +1,568 @@
+				errChan <- err
+			}
+
+			err = c.clientset.CoreV1().Pods(vmi.Namespace).Evict(context.Background(),
```
Hm, here we may have a problem. KubeVirt is actually blocking the eviction while in the background starting a migration. So while technically you get an error back, practically this succeeded. We simply have no cleaner way right now, since we can't accept that the pod gets a deletion timestamp, but on the other hand can also not return e.g. 201 while still blocking the delete.

davidvossel replied:
i can optimize for this. i won't call evict on a VMI that's currently in the process of live migrating. That will at least prevent this controller from spamming out errors as it attempts to retry Evicting

davidvossel replied:
fixed

davidvossel commented:
/retest

vladikr commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -435,6 +455,10 @@ func (c *VMIController) updateStatus(vmi *virtv1.VirtualMachineInstance, pod *k8
 			log.Log.V(3).Object(vmi).Infof("All pods have been deleted, removing finalizer")
 			controller.RemoveFinalizer(vmiCopy, virtv1.VirtualMachineInstanceFinalizer)
+			if vmiCopy.Labels != nil {
+				delete(vmiCopy.Labels, virtv1.OutdatedLauncherImageLabel)
```
I actually can't find where is this being used... :/

If it's not being used, why do we need to worry about deleting it, even if it is, seems that the vmi is deleted at this.
I think I missing something.

davidvossel replied:
The label only exists for convenience. It serves as a way for admins to identify and list all VMIs which are out of date using a label selector. I wouldn't care about this if we had a way to perform field selection for VMIs, but we don't :( 

So you're right that our controllers don't perform any action based on this label. Instead our controllers look at the VMI status's container image to determine if the VMI is out of date. But, with the label, a field engineer could easily identify out of 1000s of VMs which ones are not successfully updating. 

I originally had the label trigger our controllers as well, making the label serve both purposes. However, roman talked me out of that because the labels are mutable by end users where our VMI status is not.

> If it's not being used, why do we need to worry about deleting it, even if it is, seems that the vmi is deleted at this.
I think I missing something.

i remove the label when the VM is in a finalized state. That doesn't necessarily mean the VMI is deleted.  It could be that the VMI exited on it's own and the VMI object will still stick around for a bit. Removing the label ensures anyone filtering by this label will only see active VMIs.

vladikr replied:
Thanks for the explanation!

> i remove the label when the VM is in a finalized state. That doesn't necessarily mean the VMI is deleted. It could be that the VMI exited on it's own and the VMI object will still stick around for a bit. Removing the label ensures anyone filtering by this label will only see active VMIs.

Yea. It makes sense to me now. I didn't understand that there will be something dependent on this label.

iholder-redhat replied:
Hey @davidvossel!
Can you please update the PR description to align with what's explained here? IIUC, the following is now incorrect:

> this label is used by the automated workload update controller to detect outdated VMIs. It's also helpful for users who may want to create their own logic to handle outdated VMIs external of the automation in this PR.

vladikr commented on pkg/virt-controller/watch/vmi.go:
```diff
@@ -309,6 +309,25 @@ func conditionsEqual(a []virtv1.VirtualMachineInstanceCondition, b []virtv1.Virt
+		if vmi.Labels == nil {
+			vmi.Labels = map[string]string{}
+		}
+		vmi.Labels[virtv1.OutdatedLauncherImageLabel] = ""
+	} else {
+		if vmi.Labels != nil {
+			delete(vmi.Labels, virtv1.OutdatedLauncherImageLabel)
```
here too. I just can't find what this label triggers...

vladikr reviewed (commented):
Sorry, it took me some time to wrap my head around this.
I'll need to make another pass today. 
I just have a few minor questions, mainly, I can't find where `OutdatedLauncherImageLabel` is being used. Seems the controller doesn't react to this label...

davidvossel commented on pkg/controller/controller.go:
```diff
@@ -250,3 +250,43 @@ func ApplyVolumeRequestOnVMISpec(vmiSpec *v1.VirtualMachineInstanceSpec, request
+			curPod = pod
+		} else if curPod.CreationTimestamp.Before(&pod.CreationTimestamp) {
+			curPod = pod
+		}
```
yup, makes sense. fixed

vladikr commented on pkg/virt-controller/watch/workload-updater/workload-updater.go:
```diff
@@ -0,0 +1,571 @@
+	// the event that some are having difficulty being relocated
+	rand.Shuffle(len(data.migratableOutdatedVMIs), func(i, j int) {
+		data.migratableOutdatedVMIs[i], data.migratableOutdatedVMIs[j] = data.migratableOutdatedVMIs[j], data.migratableOutdatedVMIs[i]
+	})
```
I guess we could just form this list of `migratableOutdatedVMIs` in a way that VMIs that have a failed migration in its migrationStatus will be last, but I'm fine with his approach too

vladikr commented:
I went over everything. It looks good to me.
Thanks for the explanation @davidvossel . Really great work!

/lgtm
